### ISSUE TITLE
`&[T]!`  (addressof IUO array) fails to compile

### ISSUE THREAD

**Issue by za-creature** [apple/swift#66824]
**Motivation**
C style pointers are still a thing. Getting rid of them would be lovely, but that involves the work of a lot of people writing safe bindings around existing (supposedly) unsafe code. This report exists because said work isn't yet done.

Many number crunching functions, including BLAS / LAPACK and otherwise a big chunk of the Accelerate framework expect (possibly mutable) C pointers to `[T]` where `T` is `Int<Size>`, `UInt<Size>` or `Float<Size>` or some interleaved variation thereof. On the other end, Swift has already acknowledged the benefits of relaxing explicit `nil` checks when the developer signals that a runtime error is unlikely or ideally (if only!), impossible. This is apparent by the existence of the `!` force-unwrap operator as well as the implementation of [`Proposal #54`](https://github.com/apple/swift-evolution/blob/main/proposals/0054-abolish-iuo.md) and its historical precedents as a language feature.

The following code compiles and performs as expected:

```swift
import Accelerate


var fft = vDSP_DFT_zrop_CreateSetup(nil, 256, .FORWARD)!


func test1() {
    var re: [Float] = Array(repeating: 1, count: 128)
    var im: [Float] = Array(repeating: 1, count: 128)
    vDSP_DFT_Execute(fft, re, im, &re, &im)
    assert(abs(re[0] - 512) < .leastNormalMagnitude)
    print("test1 ok")
}
test1()


func test2() {
    var re: [Float]!
    var im: [Float]!
    re = Array(repeating: 1, count: 128)
    im = Array(repeating: 1, count: 128)

    re.withUnsafeMutableBufferPointer { let re = $0.baseAddress!
    im.withUnsafeMutableBufferPointer { let im = $0.baseAddress!
        vDSP_DFT_Execute(fft, re, im, re, im)
    }}
    assert(abs(re[0] - 512) < .leastNormalMagnitude)
    print("test2 ok")
}
test2()
```

**Solution**
This should compile and use the existing force-unwrap semantics to cast `inout [T]?` to `UnsafeMutablePointer<T>`:

```swift
// WILL NOT COMPILE: Cannot convert value of type 'UnsafeMutablePointer<[Float]?>'
//                      to expected argument type 'UnsafeMutablePointer<Float>'
func test3() {
    var re: [Float]!
    var im: [Float]!
    re = Array(repeating: 1, count: 128)
    im = Array(repeating: 1, count: 128)

    vDSP_DFT_Execute(fft, re, im, &re, &im)
    assert(abs(re[0] - 512) < .leastNormalMagnitude)
    print("test3 ok")
}
test3()
```

**Alternatives considered**
Honestly, this is just a bug. While I understand that the language is designed to prevent the programmer from going down a slippery slope in a large codebase, the fact of the matter is that attempting to access an uninitialized IUO value will result in a runtime error just as well.

There is a precedent here: `re` and `im` are already implicitly cast to `UnsafePointer<T>`^1 and defining another set of non-IUO parameters `(&re1, &im1)` for the output will run perfectly fine:

```swift
func test4() {
    var re: [Float]!
    var im: [Float]!
    re = Array(repeating: 1, count: 128)
    im = Array(repeating: 1, count: 128)

    var re1: [Float] = Array(repeating: 1, count: 128)
    var im1: [Float] = Array(repeating: 1, count: 128)

    vDSP_DFT_Execute(fft, re, im, &re1, &im1)
    assert(abs(re1[0] - 512) < .leastNormalMagnitude)
    print("test4 ok")
}
test4()
```

^1 as of whatever version of swift Xcode 14.3.1 (14E300c) is running

**Comment by za-creature** [comment]
Eh, kind-of. That (together with a comment pointing here) is what I ultimately ended up with, but then found out about implicitly unwrapped optionals and updated my code to use them instead. I opened this because the FFT was the only straggler, did some investigating and pinpointed it to what is inconsistent typecasting to `Unsafe[Mutable?]Pointer`.

I'll use the current example to describe the bigger issue that I've been dealing with. A TL;DR: might be "chicken & egg meets DRY"

So I have this thing (name pending). Outside of unit testing, it's always a singleton but I see no benefit in enforcing this in any way.

```swift
class Thing {
    // ...
}
```

My thing accepts live audio samples from the mic, so it's really:
```swift
class Thing: AVCaptureAudioDataOutputSampleBufferDelegate {
    func captureOutput(_ output: AVCaptureOutput,
                       didOutput sampleBuffer: CMSampleBuffer,
                       from connection: AVCaptureConnection) {
        // ...
    }
}
```

It also performs a FFT and delivers dB power bins to the UI, so it's also:
```swift
class Thing: ObservableObject {
    @Published var bins: [Float]
}
```

But then, the user might decide that they want more (or fewer) bins for their visualizer, so my thing has to provide a means for the UI to read and alter that:
```swift
    @Published private(set) var binCount: Int
    func updateBinCount(_ newValue: Int) {
        // ...
    }
```

For the sake of this conversation, we'll assume that everything is running on the main queue to avoid peppering the code with dispatches. In practice multiple queues are involved, so I can't really use `{ didSet { /* resize here */ } }` because I either have to dispatch sync (blocking `Main`) or dispatch async and violate the read-after-write consistency that the UI might expect from a property.

Anyway, for the fft I'll need `re_samples`  and `im_samples` (the slightly clumsy packing format used by the real-to-complex fft) each of size `binCount`, so I'd have to initialize them in the constructor:

```swift
    private var re_samples: [Float]
    private var im_samples: [Float]
    func init(binCount: Int) {
        re_samples = [Float](repeating: 0, count: binCount)
        im_samples = [Float](repeating: 0, count: binCount)
    }
}
```

But that's not the whole story. What I've really implemented [is this](https://webaudio.github.io/web-audio-api/#fft-windowing-and-smoothing-over-time) so I also need to keep track of the previous amplitudes (`Float[binCount]`) and the result of [Blackman's not-very-serious-proposal](https://en.wikipedia.org/wiki/Window_function#Blackman_window) scaled to `binCount` and (for ease of operations) split into `re_window` and `im_window`. I went a bit beyond the spec and also added noise attenuation, so the noise weighting curve ([personal preference](https://en.wikipedia.org/wiki/ITU-R_468_noise_weighting)) also needs to be recomputed not only when `binCount` changes but also whenever (if at all) the OS decides to change the mic's sampling frequency.

Anyway, the point is: a lot of work is happening inside `init` and most of that work also needs to happen whenever some `update` function is called (this is the DRY factor mentioned above). Unfortunately, the compiler will (with good reason) complain that I can't call `update()` from within `init()` because not all props have been initialized on `self` (this is the chicken-and-egg factor). I can avoid that by making everything an Optional, but then I'd have to shout! everywhere and implicitly-unwrapped-optionals are a way around that.

Setting a default value (i.e. `= []`) without changing to Optional also works, but that makes readers wonder `WTF?` whenever they encounter the double-assignment which I'd also like to avoid by adding a somewhat explicit "not initialized here but pinky-promise it'll be there when you need it" statement, which is what I see `var a: Int!` being all about.

Now one may argue that if such a big structural change is happening within the object, maybe it should be discarded and replaced with a new object initialized with different parameters. I'm somewhat against that in general because it eliminates any potential for optimization (maybe some data can be kept or some computations avoided when resizing instead of reinitializing) but strongly against it in this case because it adds a lot of unnecessary bookkeeping:
1. The object is a delegate, so it needs to be replaced upstream
2. Replacing the object entirely will probably issue a full UI update likely with `bins = [0...]` until the new object accumulates sufficient samples to perform another FFT.
3. Any other non-UI Combine subscribers will also need to re-sub to the new object.

In general, I like this pattern that I guess you could call `reconfigurable RAII` because it allows me (or the caller) to decide when to (partially) reinitialize without running the risk that the caller forgets to call some mandatory init function and ends up with a runtime error:

```swift
class Thing {
    private var a: A!
    private var b: B!
    private var c: C!
    init(a: AConfig, b: BConfig, c: CConfig) {
        updateA(a)
        updateB(b)
        updateC(c)
    }

    func updateA(_ a: AConfig) { self.a = /* ... */ }
    func updateB(_ b: BConfig) { self.b = /* ... */ }
    private func updateC(_ c: CConfig) { self.c = /* ... */ } // nope, sorry, too dangerous to expose
}

**Comment by tbkka** [comment]
If I understand correctly, your `updateA` does not depend on the existing contents of your object at all.  So it doesn't need to be a member function.  It can be a static member of `Thing` or it could be a free function outside of `Thing`.  That would give you:
```
private func recomputeA(_ a: AConfig) -> A { ... }

class Thing {
    private var a: A
    init(a: AConfig, ...) {
        self.a = recomputeA(a)
    }
    func updateA(_ a: AConfig) { self.a = recomputeA(a)}
}
```

**Comment by za-creature** [comment]
Yes, that's also a way to do it and not a bad one at that. I prefer it as a (mutating) member function because:

1. it encapsulates all behavior inside the class (or struct) itself (though the same could be said about static)
2. I get the possibility to bulk-optimize (e.g. this is a bad idea, but conceivably I could alloc a single big buffer and split it across the various arrays)
3. I could perform correlations with full private access to the object (e.g. to compute the noise curve, I don't just need the binCount but also the samplingFrequency so in general I'm left with either updating the prototype on every change, or just passing `self` to `recompute` in which case it's really a method in a C-object sort of way: `ClassName_methodName(ClassName *self, args...)` )

Basically, even though I don't necessarily use existing object data at a given time, I like to keep that option available (without refactoring) in case I may need it in the future. Whenever that's a bad idea, I feel that a strongly worded code comment describing the scenario has a higher likelihood of preventing my future self or other devs from attempting it than making it syntactically more difficult to do so (e.g. nerds love challenges, an architect's clever setup is another dev's pointer arithmetic)

**Comment by za-creature** [comment]
After giving it another thought, I believe that your most recent suggestion has enough desirable properties overall to warrant the overhead of updating function prototypes and submit to your argument.

Returning to the topic at hand, I propose that either:
1. `IUO<[T]>` may not be cast to `UnsafePointer<T>` or
2. `&IUO<[T]>` may be cast to `UnsafeMutablePointer<T>`
